File: java/1220-Count-Vowels-Permutation.java
class Solution {
    int MOD = (int) 1e9+7; 

    public int countVowelPermutation(int n) {
        if (n == 1) {
            return 5;
        }

        long[][] dp = new long[n + 1][5];
        for (int j = 0; j < 5; j++) {
            dp[1][j] = 1;
        }

        for (int i = 2; i <= n; i++) {
            dp[i][0] = dp[i - 1][1];                            
            dp[i][1] = (dp[i - 1][0] + dp[i - 1][2]) % MOD;  
            dp[i][2] = (dp[i - 1][0] + dp[i - 1][1] + dp[i - 1][3] + dp[i - 1][4]) % MOD; 
            dp[i][3] = (dp[i - 1][2] + dp[i - 1][4]) % MOD;    
            dp[i][4] = dp[i - 1][0];                           
        }

        long result = 0;
        for (int j = 0; j < 5; j++) {
            result = (result + dp[n][j]) % MOD;
        }

        return (int)result;
    }
}

File: c++/1220-count-vowels-permutation.cpp
class Solution {
const unsigned int mod = 1e9+7;
public:
    int countVowelPermutation(int n) {
        vector<int> prev(5,1), curr(5, 0);
        for(int i=1; i<n; i++) {
            curr[0] = prev[1] % mod;
            curr[1] = (prev[0] + prev[2]) % mod;
            curr[2] = ((prev[0] % mod) + (prev[1] % mod) + (prev[3] % mod) + (prev[4] % mod)) % mod;
            curr[3] = (prev[4] + prev[2]) % mod;
            curr[4] = prev[0] % mod;
            prev = curr;
        }
        int ans = 0;
        for(auto &a:prev) ans = (ans + a) % mod;
        return ans;
    }
};
